libcloudforensics/aws.py

```diff
@@ -0,0 +1,695 @@
+ACCOUNT_SERVICE = 'sts'
+REGEX_TAG_VALUE = re.compile('^.{1,255}$')
+
+
+class AWSAccount:
+  """Class representing an AWS account.
+
+  Attributes:
+    default_availability_zone (str): Default zone within the region to create
+        new resources in.
+  """
+  def __init__(self, default_availability_zone):
+    self.default_availability_zone = default_availability_zone
+    # The region is given by the zone minus the last letter
+    # https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions
+    # -availability-zones.html#using-regions-availability-zones-describe
+    self.default_region = self.default_availability_zone[:-1]
+
+  def ClientApi(self, service, region=None):
+    """Create an AWS client object.
+
+    Args:
+      service (str): The AWS service to use.
+      region (str): Optional. The region is which to create new resources in.
+
+    Returns:
+      boto3.Session.Client: An AWS EC2 client object.
+    """
+    if region:
+      return boto3.session.Session().client(
+          service_name=service, region_name=region
+      )
+    return boto3.session.Session().client(
+        service_name=service, region_name=self.default_region
+    )
+
+  def ResourceApi(self, service, region=None):
+    """Create an AWS resource object.
+
+    Args:
+      service (str): The AWS service to use.
+      region (str): Optional. The region is which to create new resources in.
+
+    Returns:
+      boto3.Session.Resource: An AWS EC2 resource object.
+    """
+    if region:
+      return boto3.session.Session().resource(
+          service_name=service, region_name=region
+      )
+    return boto3.session.Session().resource(
+        service_name=service, region_name=self.default_region
+    )
+
+  def ListInstances(self, region=None, filters=None):
+    """List instances of an AWS account.
+
+    Args:
+      region (str): Optional. The region from which to list instances.
+      filters (list(dict)): Optional. Filters for the query.
+
+    Returns:
+      dict: Dictionary with name and metadata for each instance.
+
+    Raises:
+      RuntimeError: If instances can't be listed.
+
+    Example usage:
+      ListInstances(region='us-east-1', filters=[
+          dict(Name='instance-id', Values=['some-instance-id'])])
+    """
+    if not filters:
+      filters = []
+
+    instances = {}
+    next_token = None
+    client = self.ClientApi(EC2_SERVICE, region=region)
+
+    while True:
+      try:
+        if next_token:
+          response = client.describe_instances(
+              Filters=filters, NextToken=next_token)
+        else:
+          response = client.describe_instances(Filters=filters)
+      except client.exceptions.ClientError as exception:
+        raise RuntimeError('Error: could not retrieve instances: {0:s}'.format(
+            str(exception)))
+
+      for reservation in response['Reservations']:
+        for instance in reservation['Instances']:
+          # Terminated instances are filtered out. If
```

Comment:
Are we sure we want this? Could this be a flag to ListInstance? like terminated=False

Reply:
In what use case would users want to list instances that are terminated? Terminated instances cannot be started again and usually disappear from a user's AWS account within a few hours. If you terminate an instance and use boto3's API, it still shows up in the API's response for that period of time, which was why it was being filtered out initially.

Reply:
okay, I thought "terminated" meant "stopped".
I'd still like to be able to collect these,  we've had a couple cases when after asking for someone to stop an instance, they actually deleted it in GCP

Reply:
Added :)